File: src/vodoo/helpdesk.py
"""Helpdesk domain namespace for Vodoo."""

from typing import Any

from vodoo.generated.helpdesk import GeneratedHelpdeskNamespace


class HelpdeskNamespace(GeneratedHelpdeskNamespace):
    """Namespace for ``helpdesk.ticket`` operations."""

    def create(
        self,
        name: str,
        *,
        description: str | None = None,
        partner_id: int | None = None,
        tag_ids: list[int] | None = None,
        team_id: int | None = None,
        **extra_fields: Any,
    ) -> int:
        """Create a helpdesk ticket.

        Args:
            name: Ticket title/name.
            description: Ticket description (HTML or plain text).
            partner_id: Customer partner ID.
            tag_ids: List of tag IDs to apply.
            team_id: Helpdesk team ID.
            **extra_fields: Additional fields to set on the ticket.

        Returns:
            ID of created ticket.
        """
        values = _build_ticket_values(
            name,
            description=description,
            partner_id=partner_id,
            tag_ids=tag_ids,
            team_id=team_id,
            **extra_fields,
        )
        return self._client.create(self._model, values)


# ---------------------------------------------------------------------------
# Module-level helpers (pure data transforms)
# ---------------------------------------------------------------------------


def _build_ticket_values(
    name: str,
    *,
    description: str | None = None,
    partner_id: int | None = None,
    tag_ids: list[int] | None = None,
    team_id: int | None = None,
    **extra_fields: Any,
) -> dict[str, Any]:
    """Build the values dict for helpdesk.ticket creation."""
    values: dict[str, Any] = {**extra_fields, "name": name}
    if description is not None:
        values["description"] = description
    if partner_id is not None:
        values["partner_id"] = partner_id
    if tag_ids is not None:
        values["tag_ids"] = [(6, 0, tag_ids)]
    if team_id is not None:
        values["team_id"] = team_id
    return values


# ---------------------------------------------------------------------------
# Display functions (no client needed)
# ---------------------------------------------------------------------------

File: packages/typescript/src/namespaces/helpdesk.ts
import type { OdooClientApi } from "../client-api.js";
import { Cmd } from "../commands.js";
import { GeneratedHelpdeskNamespace } from "../generated/helpdesk.js";
import type { OdooRecord } from "../types.js";
import type { ListOptions } from "./domain.js";

export interface HelpdeskRecord extends OdooRecord {
  id: number;
  name: string;
  create_date?: Date | null;
}

export interface CreateTicketOptions {
  description?: string;
  partnerId?: number;
  tagIds?: readonly number[];
  teamId?: number;
  extraFields?: Readonly<Record<string, unknown>>;
}

export class HelpdeskNamespace extends GeneratedHelpdeskNamespace {
  constructor(client: OdooClientApi) {
    super(client);
  }

  override async list(options: ListOptions = {}): Promise<HelpdeskRecord[]> {
    return (await super.list(options)) as HelpdeskRecord[];
  }

  override async get(
    recordId: number,
    fields?: readonly string[],
  ): Promise<HelpdeskRecord> {
    return (await super.get(recordId, fields)) as HelpdeskRecord;
  }

  create(name: string, options: CreateTicketOptions = {}): Promise<number> {
    const values: Record<string, unknown> = {
      ...options.extraFields,
      name,
    };
    if (options.description !== undefined) {
      values.description = options.description;
    }
    if (options.partnerId !== undefined) values.partner_id = options.partnerId;
    if (options.tagIds !== undefined)
      values.tag_ids = [Cmd.set(options.tagIds)];
    if (options.teamId !== undefined) values.team_id = options.teamId;
    return this.client.create("helpdesk.ticket", values);
  }
}
